feat(stdlib): add unflatten vrl function

jorgehermo9 opened the pull request:
Closes #81 

Had no time to polish this PR, but I think it is good enough to be reviewed.I will do the things I've remaining asap, but pushing this if happens that someone can manage to review it in the meanwhile (no hurry)

Things I've left but couldn't do yet because had no time
- changelog
- better examples
- documentation of this function in a Vector repo PR
- Add tests so we can check that `unflatten` is the inverse of `flatten` in the tests under `lib/tests/tests/functions`

If I miss anything, tell me please

I will address this as soon as I can, but I will be off these days.

jorgehermo9 committed:
feat: implement unflatten vrl function

jorgehermo9 committed:
fix: clippy lints

jorgehermo9 commented on Cargo.toml:
```diff
@@ -142,7 +142,7 @@ iana-time-zone = { version = "0.1", optional = true }
 indexmap = { version = "~2.4.0", default-features = false, features = ["std"], optional = true}
 indoc = {version = "2", optional = true }
-itertools = { version = "0.13", default-features = false, optional = true }
+itertools = { version = "0.13", default-features = false, features=["use_std"], optional = true }
```
Need that feature to use `into_grouping_by`

jorgehermo9 committed:
fix: wrong input type

jorgehermo9 commented on src/stdlib/unflatten.rs:
```diff
@@ -0,0 +1,279 @@
+    }
+}
+
+// this should return the key to insert?
```
```suggestion
```

jorgehermo9 committed:
Update unflatten.rs

jorgehermo9 committed:
Update mod.rs

jszwedko reviewed (commented):
You are on a roll :)

This looks like it is heading in the right direction to me. I appreciate the number of tests that were added.

If you are feeling inspired, this could be a good use-case for property testing where random inputs are fed into `flatten` through `unflatten` and then asserted to be the same. This would be the first time we have these sorts of tests in the stdlib, though, so might not be straight-forward; definitely not a requirement for merging this. There are some prop tests declared for the compiler here: https://github.com/vectordotdev/vrl/tree/main/lib/proptests

jorgehermo9 committed:
test(unflatten): add integration tests

jorgehermo9 committed:
docs(stdlib): add changelog

jorgehermo9 commented on lib/tests/tests/functions/flatten/simple.vrl:
```diff
@@ -0,0 +1,4 @@
+# object: { "a": { "b": { "c": { "d": 1, "e": 2 }, "f": 3 }, "g": 4 } }
+# result: { "a.b.c.d": 1, "a.b.c.e": 2, "a.b.f": 3, "a.g": 4 }
+
+flatten(.)
```
Maybe this test is redudant with the ones from the examples
https://github.com/vectordotdev/vrl/blob/d3a7dd7fda91f6d200ec3d09b70bc20c5af4867f/src/stdlib/unflatten.rs#L96

What do you think? Should I only leave the `unflatten(flatten(.))` one?

jszwedko replied:
I think it is still a useful test to assert the unflattened representation. One could imagine a bug that causes round tripping through `flatten` to pass while the output of the `unflatten` call is incorrect.

jorgehermo9 committed:
feat(stdlib): add unflatten examples

jorgehermo9 commented:
I added a new `recursive` parameter to control wether the values should be unflattened recursively too. I hope the examples explains how it behaves. What do you think about this?

jszwedko commented on lib/tests/tests/functions/flatten/from_unflatten.vrl:
```diff
@@ -0,0 +1,4 @@
+# object: { "a.b.c.d": 1, "a.b.c.e": 2, "a.b.f": 3, "a.g": 4 }
+# result: { "a.b.c.d": 1, "a.b.c.e": 2, "a.b.f": 3, "a.g": 4 }
+
+flatten(unflatten(.))
```
👍 I like this test.

jszwedko commented on src/stdlib/unflatten.rs:
```diff
@@ -0,0 +1,329 @@
+    grouped
+        .into_iter()
+        .map(|(key, mut values)| {
+            if values.len() == 1 {
+                match values.pop().unwrap() {
```
```suggestion
            if values.len() == 1 {
                match values.pop().expect("exactly one element") {
```

We typically `expect` over `unwrap` to provide more context about why the panic should never be hit.

I think one potential way to refactor this that might avoid this unwrap/expect and avoid the early returns is something like:

```rust
match values {
  [value] => { ... what you currently have for handling if there is only one value ... }
  values => { ... handle multiple values ... }
}
```

jorgehermo9 replied:
I tried to do that, but the problem is `values`  is a `Vec` and therefore, we should convert it to a slice in order to pattern match against it, using `values[..]`, but then, at the first branch we would have the first element borrowed and we would need it owned. The same with the second branch, we will have a borrowed slice of values, but not the owned values and we need owned values for the `do_unflatten_entries` function.

Does this makes sense?

jszwedko replied:
Mmm, yeah, that makes sense. I wish there was a better way to model this that doesn't require the `expect` but I'm not seeing it.

jszwedko commented on src/stdlib/unflatten.rs:
```diff
@@ -0,0 +1,329 @@
+        let separator = arguments
+            .optional("separator")
+            .unwrap_or_else(|| expr!(DEFAULT_SEPARATOR));
+        let recursive = arguments
+            .optional("recursive")
+            .unwrap_or_else(|| expr!(false));
```
Should the default be `true` to match `flatten`?

jszwedko reviewed (commented):
Nice, thanks @jorgehermo9 ! I left a few more comments, but this is looking good to me. No worries about the prop tests. I appreciate you giving those a shot, but definitely not required for this.

jorgehermo9 committed:
feat(stdlib): improve unflatten function with single keys

jorgehermo9 commented:
Hi @jszwedko 

I addressed the comments. I additionally add a more efficient way of constructing objects with "unique paths", for example:
```json
{
  "a.b.c.d.e.f": 1,
  "g.h.i.j.k" 2,
}
```
Since the two entries don't share any prefix, they can be constructed without the recursive calls to `do_unflatten_entries` and then we can construct the objects iteratively with the `do_unflatten_entry`. 

This may be a premature optimization, but seems logical to me to not construct a HashMap for grouping keys recursively if there will be just one key group at each recursive call (see https://github.com/vectordotdev/vrl/pull/993/commits/6a0758ddb0778292e21785264a9a88f2ba667a62)

jorgehermo9 committed:
test(stdlib): add test for consecutive separators in unflatten function

jorgehermo9 committed:
test(stdlib): add test for consecutive separators in unflatten function

jszwedko reviewed (approved):
Great! Thanks for this contribution @jorgehermo9 !

jorgehermo9 committed:
fix: clippy lints

jorgehermo9 committed:
Merge with main

jorgehermo9 commented:
Sorry, clippy action failed in CI. I though that `script/checks.sh` would run `cargo clippy` but I see that it doesn't.

Could you approve again? @jszwedko

jorgehermo9 committed:
docs: include authors in changelog
